src/proxy_configuration.js

```diff
@@ -0,0 +1,319 @@
+            } else {
+                this.password = password;
+            }
+        } else if (!this.password) {
+            throw new Error(this._getMissingParamErrorMgs('password', ENV_VARS.PROXY_PASSWORD));
+        }
+    }
+
+    /**
+     *  Checks the status of Apify Proxy and throws an error if the status is not "connected".
+     * @returns {Promise<void>}
+     * @ignore
+     */
+    async _checkAccess() {
+        const url = APIFY_PROXY_STATUS_URL;
+        const proxyUrl = this.getUrl();
+        const { countryCode } = this;
+        const { body: { connected, connectionError } } = await requestAsBrowser({ url, proxyUrl, countryCode, json: true });
+        if (!connected) this._throwApifyProxyConnectionError(connectionError);
+    }
+
+    /**
+     *  Validates if parameters groups and countryCode have correct structure
+     * @ignore
+     */
+    _validateArgumentStructure(groups, countryCode) {
+        if (groups && groups.length) {
+            for (const group of groups) {
+                if (!APIFY_PROXY_VALUE_REGEX.test(group)) this._throwInvalidProxyValueError(group);
+            }
+        }
+        if (countryCode) {
+            if (!COUNTRY_CODE_REGEX.test(countryCode)) this._throwInvalidCountryCode(countryCode);
+        }
+    }
+
+    /**
+     *  Returns missing parameter error message.
+     * @param {string} param
+     * @param {string} env
+     * @return {string} - error message
+     * @ignore
+     */
+    _getMissingParamErrorMgs(param, env) {
+        return `Apify Proxy ${param} must be provided as parameter or "${env}" environment variable!`;
+    }
+
+    /**
+     *  Throws invalid proxy value error
+     * @param {string} param
+     * @ignore
+     */
+    _throwInvalidProxyValueError(param) {
+        throw new Error(`The provided proxy group name "${param}" can only contain the following characters: 0-9, a-z, A-Z, ".", "_" and "~"`);
+    }
+
+    /**
+     *  Throws invalid country code error
+     * @param {string} code
+     * @ignore
+     */
+    _throwInvalidCountryCode(code) {
+        throw new Error(`The provided country code "${code}" is not valid. Please use a two letter country code according to ISO 3166-1 alpha-2`);
+    }
+
+    /**
+     *  Throws Apify Proxy is not connected
+     * @ignore
+     */
+    _throwApifyProxyConnectionError(errorMessage) {
+        throw new Error(errorMessage);
+    }
+}
+
+/**
+ *  Creates a proxy configuration and returns a promise resolving to an instance
+ *  of the {@link ProxyConfiguration} class that is already initialized.
+ *
+ *  Factory function which gets the {ProxyConfigurationOptions} proxyConfigurationOptions as a parameter and returns
+ *  preconfigured {@link ProxyConfiguration} with already passed all check and validations so it should  be save
+ *  to use it as a configuration for your crawler and call its methods.
```

Comment:
This feels that it says the same thing as the previous paragraph, only in a more confusing way. See the other similar functions such as `openRequestList` or `openKeyValueStore` for inspiration.

Reply:
Changed the description.